File: src/modules/auth/auth.controller.ts
import { Controller, Post, Req, Body, UseGuards } from "@nestjs/common";
import { AuthGuard } from "@nestjs/passport";
import { AuthService } from "./auth.service";
import {
  ApiTags,
  ApiOperation,
  ApiResponse,
  ApiBody,
  ApiBearerAuth,
  ApiSecurity,
} from "@nestjs/swagger";
import { LoginDto } from "./dto/login.dto";
import { GoogleLoginDto } from "./dto/google-login.dto";
import { AppleLoginDto } from "./dto/apple-login.dto";
import { UserRole } from "src/enums/user-role.enum";
import { SignupDto } from "./dto/signup.dto";
import { RegisterInitDto } from "./dto/register-init.dto";
import { ForgotPasswordDto } from "./dto/forgot-password.dto";
import { ResetPasswordDto } from "./dto/reset-password.dto";

@ApiTags("auth")
@Controller("auth")
export class AuthController {
  constructor(private readonly authService: AuthService) {}

  @Post("google/login")
  @ApiOperation({ summary: "Login via Google using ID Token" })
  @ApiBody({ type: GoogleLoginDto })
  @ApiResponse({
    status: 200,
    description: "Returns JWT access token after validating Google ID Token",
  })
  async googleLoginWithToken(@Body() dto: GoogleLoginDto) {
    return this.authService.googleLogin(dto.idToken);
  }

  @Post("apple/login")
  @ApiOperation({ summary: "Login via Apple using ID Token" })
  @ApiBody({ type: AppleLoginDto })
  @ApiResponse({
    status: 200,
    description: "Returns JWT access token after validating Apple ID Token",
  })
  async appleLoginWithToken(@Body() dto: AppleLoginDto) {
    return this.authService.appleLogin(dto.idToken);
  }

  @Post("login")
  @UseGuards(AuthGuard("email"))
  @ApiOperation({
    summary: "Local login with email/password",
  })
  @ApiBody({ type: LoginDto })
  @ApiResponse({
    status: 200,
    description:
      "Returns JWT access token on successful login with available roles",
  })
  async login(@Req() req) {
    return this.authService.login(req.user);
  }

  @Post("switch-role")
  @UseGuards(AuthGuard("jwt"))
  @ApiBearerAuth()
  @ApiSecurity("jwt")
  @ApiOperation({ summary: "Switch active role and receive new JWT token" })
  @ApiBody({
    schema: {
      type: "object",
      properties: {
        role: {
          type: "string",
          enum: Object.values(UserRole),
          example: "admin",
        },
      },
      required: ["role"],
    },
  })
  @ApiResponse({
    status: 200,
    description: "Returns a new JWT access token with the selected role active",
  })
  async switchRole(@Req() req, @Body("role") role: UserRole) {
    return this.authService.switchActiveRole(req.user, role);
  }

  @Post("register-init")
  @ApiOperation({ summary: "Inicia o processo de registro enviando email" })
  @ApiBody({ type: RegisterInitDto })
  @ApiResponse({
    status: 200,
    description: "Email de registro enviado com sucesso",
  })
  @ApiResponse({ status: 400, description: "Email já cadastrado" })
  async registerInit(@Body() registerInitDto: RegisterInitDto) {
    return this.authService.registerInit(registerInitDto);
  }

  @Post("signup")
  @ApiOperation({ summary: "Completa o registro de um novo usuário com token" })
  @ApiResponse({
    status: 201,
    description: "Usuário registrado com sucesso",
  })
  @ApiResponse({
    status: 400,
    description: "Dados inválidos ou email já existe",
  })
  @ApiResponse({
    status: 401,
    description: "Token de registro inválido ou expirado",
  })
  async signup(@Body() signupDto: SignupDto) {
    return this.authService.signup(signupDto);
  }

  @Post("forgot-password")
  @ApiOperation({ summary: "Envia um email de recuperação de senha" })
  @ApiBody({ type: ForgotPasswordDto })
  @ApiResponse({
    status: 200,
    description: "Email de recuperação de senha enviado com sucesso",
  })
  @ApiResponse({ status: 404, description: "Usuário não encontrado" })
  async forgotPassword(@Body() forgotPasswordDto: ForgotPasswordDto) {
    return this.authService.forgotPassword(forgotPasswordDto.email);
  }

  @Post("reset-password")
  @ApiOperation({ summary: "Reseta a senha do usuário" })
  @ApiBody({ type: ResetPasswordDto })
  @ApiResponse({ status: 200, description: "Senha resetada com sucesso" })
  @ApiResponse({
    status: 400,
    description: "Dados inválidos ou token inválido",
  })
  async resetPassword(@Body() resetPasswordDto: ResetPasswordDto) {
    return this.authService.resetPassword(resetPasswordDto);
  }
}

File: RapidaQuickstart.DotNet/Controllers/AuthController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using RapidaQuickstart.DotNet.Common.Services;
using RapidaQuickstart.DotNet.DTOs;
using RapidaQuickstart.DotNet.Enums;
using RapidaQuickstart.DotNet.Models;
using RapidaQuickstart.DotNet.Services;
using System.Security.Claims;

namespace RapidaQuickstart.DotNet.Controllers
{
    /// <summary>
    /// Authentication and authorization endpoints
    /// </summary>
    [ApiController]
    [Route("api/auth")]
    public class AuthController : ControllerBase
    {
        private readonly IAuthService _authService;
        private readonly IUserService _userService;
        private readonly IErrorService _errorService;

        public AuthController(IAuthService authService, IUserService userService, IErrorService errorService)
        {
            _authService = authService;
            _userService = userService;
            _errorService = errorService;
        }

        /// <summary>
        /// Login via Google using ID Token
        /// </summary>
        /// <param name="dto">Google login credentials</param>
        /// <returns>JWT access token and user information</returns>
        /// <response code="200">Returns JWT access token after validating Google ID Token</response>
        /// <response code="400">Invalid Google ID Token</response>
        [HttpPost("google/login")]
        public async Task<IActionResult> GoogleLogin([FromBody] GoogleLoginDto dto)
        {
            try
            {
                var result = await _authService.GoogleLoginAsync(dto.IdToken);
                return Ok(result);
            }
            catch (Exception ex)
            {
                return BadRequest(new { message = ex.Message });
            }
        }

        [HttpPost("apple/login")]
        public async Task<IActionResult> AppleLogin([FromBody] AppleLoginDto dto)
        {
            try
            {
                var result = await _authService.AppleLoginAsync(dto.IdToken);
                return Ok(result);
            }
            catch (Exception ex)
            {
                return BadRequest(new { message = ex.Message });
            }
        }

        /// <summary>
        /// Local login with email and password
        /// </summary>
        /// <param name="dto">Login credentials</param>
        /// <returns>JWT access token and user information</returns>
        /// <response code="200">Returns JWT access token on successful login</response>
        /// <response code="401">Invalid credentials</response>
        /// <response code="400">Invalid request data</response>
        [HttpPost("login")]
        public async Task<IActionResult> Login([FromBody] LoginDto dto)
        {
            try
            {
                var result = await _authService.LoginAsync(dto);
                return Ok(result);
            }
            catch (UnauthorizedAccessException ex)
            {
                return Unauthorized(new { message = ex.Message });
            }
            catch (Exception ex)
            {
                return BadRequest(new { message = ex.Message });
            }
        }

        /// <summary>
        /// Switch active role and receive new JWT token
        /// </summary>
        /// <param name="dto">Role to switch to</param>
        /// <returns>New JWT access token with selected role active</returns>
        /// <response code="200">Returns new JWT access token with selected role active</response>
        /// <response code="401">Unauthorized or user doesn't have the requested role</response>
        /// <response code="400">Invalid request data</response>
        [HttpPost("switch-role")]
        [Authorize]
        public async Task<IActionResult> SwitchRole([FromBody] SwitchRoleDto dto)
        {
            try
            {
                var userId = User.FindFirst("userId")?.Value;
                if (string.IsNullOrEmpty(userId))
                {
                    return Unauthorized(new { message = _errorService.GetErrorMessage(ErrorCode.UNAUTHORIZED) });
                }

                var user = await _userService.GetUserByIdAsync(userId);
                if (user == null)
                {
                    return NotFound(new { message = _errorService.GetErrorMessage(ErrorCode.USER_NOT_FOUND) });
                }

                var result = await _authService.SwitchActiveRoleAsync(user, dto.Role);
                return Ok(result);
            }
            catch (Exception ex)
            {
                return BadRequest(new { message = ex.Message });
            }
        }

        [HttpPost("register-init")]
        public async Task<IActionResult> RegisterInit([FromBody] RegisterInitDto dto)
        {
            try
            {
                var result = await _authService.RegisterInitAsync(dto);
                return Ok(result);
            }
            catch (InvalidOperationException ex)
            {
                return BadRequest(new { message = ex.Message });
            }
            catch (Exception ex)
            {
                return BadRequest(new { message = ex.Message });
            }
        }

        [HttpPost("signup")]
        public async Task<IActionResult> Signup([FromBody] SignupDto dto)
        {
            try
            {
                var result = await _authService.SignupAsync(dto);
                return Ok(result);
            }
            catch (UnauthorizedAccessException ex)
            {
                return Unauthorized(new { message = ex.Message });
            }
            catch (Exception ex)
            {
                return BadRequest(new { message = ex.Message });
            }
        }

        [HttpPost("forgot-password")]
        public async Task<IActionResult> ForgotPassword([FromBody] ForgotPasswordDto dto)
        {
            try
            {
                var result = await _authService.ForgotPasswordAsync(dto.Email);
                return Ok(result);
            }
            catch (Exception ex)
            {
                return BadRequest(new { message = ex.Message });
            }
        }

        [HttpPost("reset-password")]
        public async Task<IActionResult> ResetPassword([FromBody] ResetPasswordDto dto)
        {
            try
            {
                var result = await _authService.ResetPasswordAsync(dto);
                return Ok(result);
            }
            catch (UnauthorizedAccessException ex)
            {
                return Unauthorized(new { message = ex.Message });
            }
            catch (Exception ex)
            {
                return BadRequest(new { message = ex.Message });
            }
        }
    }
}
